### PR TITLE
Fix external statistics initialization

### PR DESCRIPTION
**Pull request by pranjalssh**
Previously we created stats calculator before we installed any plugins - so external stats providers were never used. Fixed that by delaying stats calculator creation till end. Added a test as well

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsManager.java [hunk 1]
```diff
@@ -41,9 +51,32 @@ public void addExternalPlanStatisticsProviderFactory(ExternalPlanStatisticsProvi
...
-    public HistoryBasedPlanStatisticsCalculator getHistoryBasedPlanStatisticsCalculator(StatsCalculator delegate)
+    public void setDelegateStatsCalculator(StatsCalculator statsCalculator)
     {
-        return new HistoryBasedPlanStatisticsCalculator(
-                externalPlanStatisticsProviderFactory.map(factory -> factory.getExternalPlanStatisticsProvider()), metadata, delegate);
+        if (delegate.isPresent()) {
+            throw new IllegalStateException("delegate stats calculator can only be set once");
+        }
+        this.delegate = Optional.of(requireNonNull(statsCalculator, "stats calculator is null"));
+    }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsManager.java:60]: From the dependency injection perspective, which module are you waiting for? Is it `delegate` or something else?

File: presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsManager.java [hunk 2]
```diff
@@ -41,9 +51,32 @@ public void addExternalPlanStatisticsProviderFactory(ExternalPlanStatisticsProvi
         this.externalPlanStatisticsProviderFactory = Optional.of(externalPlanStatisticsProviderFactory);
     }
 
-    public HistoryBasedPlanStatisticsCalculator getHistoryBasedPlanStatisticsCalculator(StatsCalculator delegate)
+    public void setDelegateStatsCalculator(StatsCalculator statsCalculator)
     {
-        return new HistoryBasedPlanStatisticsCalculator(
-                externalPlanStatisticsProviderFactory.map(factory -> factory.getExternalPlanStatisticsProvider()), metadata, delegate);
```
[the remark this answers is not in the document]
Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsManager.java:47]: #12994 probably is a similar fix as an example

File: presto-main/src/test/java/com/facebook/presto/cost/TestExternalStatsProvider.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    @Test
+    public void testExternalStatsCalculator()
+    {
+        // Overridden stats
+        assertPlan("SELECT orderstatus FROM orders",
+                anyTree(node(TableScanNode.class).withOutputRowCount(100)));
+
+        // Original stats
+        assertPlan("SELECT * FROM nation",
+                anyTree(node(TableScanNode.class).withOutputRowCount(25)));
```
[the remark this answers is not in the document]
Comment (pranjalssh) [reply presto-main/src/test/java/com/facebook/presto/cost/TestExternalStatsProvider.java]: Copied from another test which has same mistakes 🤡

File: presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsManager.java:47]: Found it weird too, thanks for the suggestion - very clean now!

File: presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java [hunk 4]
```diff
@@ -36,11 +37,11 @@
 public class HistoryBasedPlanStatisticsCalculator
         implements StatsCalculator
 {
-    private final Optional<ExternalPlanStatisticsProvider> externalPlanStatisticsProvider;
+    private final Supplier<Optional<ExternalPlanStatisticsProvider>> externalPlanStatisticsProvider;
```
[the remark this answers is not in the document]
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java]: Added EmptyPlanStatisticsProvider, want to keep useExternalPlanStatisticsEnabled(session) outside together with with other checks, as its not clear right now how we want to mix these stats with history based stats  - so want to keep everything in one place.

**Review by highker — APPROVED** [review]

**Pull request merged**